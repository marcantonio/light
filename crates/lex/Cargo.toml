[package]
name = "lex"
version = "0.0.0"
edition = "2021"

[dependencies]
itertools = "0.10.5"
serde = { version = "1.0.136", features = ["derive"] }

common = { path = "../common" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }
