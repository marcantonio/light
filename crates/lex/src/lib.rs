use serde::Serialize;
use std::iter::Peekable;

use common::Operator;
pub use token::{Token, TokenType};

#[cfg(test)]
mod tests;
pub mod token;

pub type LexResult = std::result::Result<Token, LexError>;

pub struct Lex {
    stream: Peekable<StreamIter<char>>,
    pub tokens: Vec<Token>,
}

impl Lex {
    pub fn new(input: &str) -> Self {
        Lex { stream: StreamIter::new(input).peekable(), tokens: vec![] }
    }

    // Scan all input
    pub fn scan(mut self) -> Result<Vec<Token>, LexError> {
        loop {
            let token = self.lex()?;
            if token.is_eof() {
                break;
            }
            self.tokens.push(token);
        }
        Ok(self.tokens)
    }

    // Recursively process enough characters to produce one token
    fn lex(&mut self) -> LexResult {
        use TokenType::*;

        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => unreachable!("can't lex nothing"),
        };

        // Inject a semicolon if certain tokens occur at the end of the line or
        // EOF. If EOF, make sure the context is right.
        if cur.value == '\n' && self.should_add_semicolon() {
            return Ok(Token::new(Semicolon(true), cur.line, cur.column));
        } else if cur.is_eof() {
            if self.should_add_semicolon() {
                let semi = match self.tokens.last() {
                    Some(t) => Token::new(Semicolon(true), t.line, t.column + 1),
                    None => Token::default(),
                };
                self.tokens.push(semi);
            }
            return Ok(Token::new(Eof, cur.line, cur.column));
        }

        // Ignore whitespace
        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Single line comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Keywords, types, and identifiers
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "fn" => Fn,
                "let" => Let,
                "for" => For,
                "if" => If,
                "else" => Else,
                "extern" => Extern,
                "true" => Bool(true),
                "false" => Bool(false),
                "struct" => Struct,
                "module" => Module,
                _ => Ident(identifier),
            };

            return Ok(Token::new(tt, cur.line, cur.column));
        }

        // Literal numbers
        if cur.value.is_ascii_digit() {
            let mut n = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    n.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return Ok(Token::new(Num(n), cur.line, cur.column));
        }

        // Literal char
        if cur == '\'' {
            let mut ch = String::new();
            let next = self.stream.next().unwrap_or_else(|| unreachable!("lexed None when looking for char"));

            match next.value {
                // Control characters
                '\\' => {
                    if let Some(next) = self.stream.next() {
                        match next.value {
                            'n' => ch = String::from("\n"),
                            't' => ch = String::from("\t"),
                            '\'' => ch = String::from("'"),
                            c => {
                                return Err(LexError::from((
                                    format!("Invalid character control sequence: `\\{}`", c),
                                    next,
                                )))
                            },
                        }
                    }
                },
                // EOF
                '\0' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`, got `EOF`".to_string(),
                        cur,
                    )));
                },
                '\'' => return Err(LexError::from(("Character literal can't be empty".to_string(), cur))),

                // Everything else
                c => ch = String::from(c),
            }

            // Check for closing '\''
            let last =
                self.stream.next().unwrap_or_else(|| unreachable!("fatal: lexed None when looking for `'`"));
            match last.value {
                '\'' => (),
                '\0' | '\n' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`".to_string(),
                        last,
                    )));
                },
                _ => {
                    return Err(LexError::from((
                        format!("Invalid character sequence: `'{}{}'`", ch, last.value),
                        last,
                    )));
                },
            }

            return Ok(Token::new(Char(ch), cur.line, cur.column));
        }

        // Multi-character operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Eq), cur.line, cur.column));
                },
                '!' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::NotEq), cur.line, cur.column));
                },
                '>' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::GtEq), cur.line, cur.column));
                },
                '<' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::LtEq), cur.line, cur.column));
                },
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::And), cur.line, cur.column));
                },
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Or), cur.line, cur.column));
                },
                '+' if next == &'+' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Inc), cur.line, cur.column));
                },
                '+' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::AddEq), cur.line, cur.column));
                },
                '-' if next == &'-' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Dec), cur.line, cur.column));
                },
                '-' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::SubEq), cur.line, cur.column));
                },
                '-' if next == &'>' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::RetType), cur.line, cur.column));
                },
                '*' if next == &'*' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Pow), cur.line, cur.column));
                },
                '*' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::MulEq), cur.line, cur.column));
                },
                '/' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::DivEq), cur.line, cur.column));
                },
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => Op(Operator::Add),
            '-' => Op(Operator::Sub),
            '*' => Op(Operator::Mul),
            '/' => Op(Operator::Div),
            '>' => Op(Operator::Gt),
            '<' => Op(Operator::Lt),
            '!' => Op(Operator::Not),
            '=' => Op(Operator::Assign),
            '&' => Op(Operator::BitAnd),
            '^' => Op(Operator::BitXor),
            '|' => Op(Operator::BitOr),
            '.' => Dot,
            '}' => CloseBrace,
            ']' => CloseBracket,
            ')' => CloseParen,
            ':' => Colon,
            ',' => Comma,
            '{' => OpenBrace,
            '[' => OpenBracket,
            '(' => OpenParen,
            ';' => Semicolon(false),
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            },
        };

        Ok(Token::new(tt, cur.line, cur.column))
    }

    // Add a semicolon for these tokens
    fn should_add_semicolon(&self) -> bool {
        use TokenType::*;

        if let Some(t) = self.tokens.last() {
            matches!(
                t.tt,
                Bool(_)
                    | Char(_)
                    | CloseBrace
                    | CloseParen
                    | CloseBracket
                    | Ident(_)
                    | Num(_)
                    | Op(Operator::Inc)
                    | Op(Operator::Dec)
            )
        } else {
            false
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, Clone, Copy, PartialEq)]
struct ContextElement<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextElement<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextElement { value, line: line + 1, column: column + 1 }
    }
}

impl ContextElement<char> {
    // Will cause lexing to stop if there's a null byte in the file
    fn is_eof(&self) -> bool {
        self.value == 0 as char
    }
}

impl PartialEq<char> for ContextElement<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextElement<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let opt = self.lines.get(self.line);
        let line = match opt {
            Some(l) => l,
            None => return Some(ContextElement::new(0 as char, self.line, self.column - 1)),
        };
        let cc =
            line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column)).or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc.or_else(|| Some(ContextElement::new(0 as char, self.line, self.column - 1)))
    }
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Lexing error: {} at {}:{}", self.message, self.line, self.column)
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextElement<T>)> for LexError {
    fn from((msg, cp): (String, ContextElement<T>)) -> Self {
        LexError { message: msg, line: cp.line, column: cp.column }
    }
}
