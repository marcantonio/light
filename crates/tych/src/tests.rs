use super::*;
use lex::Lex;
use parse::Parse;

const ERR_STR: &str = "Numeric literal out of range";

macro_rules! run_insta {
    ($prefix:expr, $tests:expr) => {
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for test in $tests {
                let tokens = Lex::new(test[1]).scan().expect("lexing failed in `tych` tests");
                let (ast, mut symbol_table, _, _) = Parse::new(&tokens).parse().expect("parsing failed in `tych` tests");
                let res = Tych::new(&mut symbol_table).walk(ast);
                insta::assert_yaml_snapshot!(format!("{}_{}", $prefix, test[0]), (test[1], res));
            }
        })
    };
}

#[test]
fn test_binop() {
    let tests = [
        [
            "basic",
            r#"
fn foo() {
    1 + 2
}
"#,
        ],
        [
            "basic_var",
            r#"
fn foo() {
    let a: int = 1
    a + 2
}
"#,
        ],
        [
            "complex",
            r#"
fn foo() {
    let a: int = 1
    a + 2 / 4 + (2 * 2)
}
"#,
        ],
        [
            "func_call",
            r#"
fn plusOne(x: int) -> int {
    x + 1
}

fn main() {
    1 + plusOne(3)
}
"#,
        ],
        [
            "mismatch",
            r#"
fn foo() {
    2.0 + 1
}
"#,
        ],
        [
            "bool_ret",
            r#"
fn foo() -> bool {
    2 == 1
}
"#,
        ],
        [
            "bad_lhs_assign",
            r#"
fn main(x: int) {
    x + 1 = 1
}
"#,
        ],
        [
            "bools_cmp_1",
            r#"
fn main(x: int) {
    true || false
}
"#,
        ],
        [
            "bools_cmp_2",
            r#"
fn main(x: bool) {
    x == false
}
"#,
        ],
        [
            "bad_lit_int",
            r#"
fn main(x: int) {
    true && 1
}
"#,
        ],
        [
            "bad_lit_float",
            r#"
fn main(x: int) {
    1.0 || true
}
"#,
        ],
        [
            "bad_cmp_type",
            r#"
fn main(x: int) {
    true > true
}
"#,
        ],
        [
            "bad_math_types",
            r#"
fn main(x: int) {
    true + false
}
"#,
        ],
        [
            "lit_char_1",
            r#"
fn main(x: int) {
    'a' > 'c'
}
"#,
        ],
        [
            "lit_char_2",
            r#"
fn main(x: char) {
    x > 'c'
}
"#,
        ],
    ];
    run_insta!("binop", tests);
}

#[test]
fn test_block() {
    let tests = [
        [
            "let",
            r#"
let x: int = {
    1
}
"#,
        ],
        [
            "let_mismatch_1",
            r#"
let x: int = {
    1.0
}
"#,
        ],
        [
            "let_mismatch_2",
            r#"
fn foo() {
    let x: int = {
        11.0
        foo()
    }
}
"#,
        ],
    ];
    run_insta!("block", tests);
}

#[test]
fn test_call() {
    let tests = [
        [
            "basic",
            r#"
fn plusOne(x: int) -> int {
    x + 1
}
fn main() {
    plusOne(1)
}
"#,
        ],
        [
            "mismatch_arg_type",
            r#"
fn plusOne(x: int) -> int {
    x + 1
}
fn main() {
    let x: float = 1.0
    plusOne(x)
}
"#,
        ],
        [
            "mismatch_arg_num",
            r#"
fn plusOne(x: int) -> int {
    x + 1
}
fn main() {
    plusOne(1, 2)
}
"#,
        ],
        [
            "undef_func",
            r#"
fn plusOne(x: int) -> int {
    x + 1
}
fn main() {
    plusThree(1)
}
"#,
        ],
    ];
    run_insta!("call", tests);
}

#[test]
fn test_cond() {
    let tests = [
        [
            "basic",
            r#"
fn foo() {
    if 1 > 2 {
        1
    } else {
        2
    }
}
"#,
        ],
        [
            "bad_cond_1",
            r#"
fn foo() {
    let x: int8 = 1
    let y: int16 = 2
    if x < y {
        1
    }
}
"#,
        ],
        [
            "bad_cond_2",
            r#"
fn foo() {
    if 1.0 {
        1
    }
}
"#,
        ],
        [
            "mismatch_arms",
            r#"
fn foo() {
    if 1 > 2 {
        1
    } else {
        2.0
    }
}
"#,
        ],
    ];
    run_insta!("cond", tests);
}

#[test]
fn test_def_func() {
    let tests = [
        [
            "basic",
            r#"
fn foo() {
    1
}
"#,
        ],
        [
            "void_ret_void_body",
            r#"
fn bar() { }
fn foo() {
    bar()
}
"#,
        ],
        [
            "void_ret_expr_body",
            r#"
fn foo(a: int) {
    a + 1
}
"#,
        ],
        [
            "multi_args",
            r#"
fn foo(a: int, b: float) { }
"#,
        ],
        [
            "expected_int_last_stmt_1",
            r#"
fn foo(a: int, b: float) -> int { }
"#,
        ],
        [
            "expected_int_last_stmt_2",
            r#"
fn foo(a: int, b: float) -> int {
    1.0
}
"#,
        ],
        [
            "extern",
            r#"
extern fn foo()
"#,
        ],
        [
            "stmt_end_of_body",
            r#"
fn foo() {
    let a: int8 = 1
}
"#,
        ],
        [
            "main_void",
            r#"
fn main() { }
"#,
        ],
        [
            "main_antn",
            r#"
fn main() -> int { }
"#,
        ],
    ];
    run_insta!("def_func", tests);
}

#[test]
fn test_for() {
    let tests = [
        [
            "basic",
            r#"
fn foo() {
    for x: int8 = 1; x < 2; 1 {
        x
    }
}
"#,
        ],
        [
            "for_shadowing1",
            r#"
fn foo() {
    let x: int32 = 1
    for x: int8 = 1; x < 2; 1 {
        x
    }
    x
}
"#,
        ],
        [
            "for_shadowing2",
            r#"
fn foo() {
    let x: [int32; 3] = [1, 2, 3]
    x[0]
    for x: int8 = 1; x < 2; 1 {
        x
    }
    x[0]
}
"#,
        ],
        [
            "for_shadowing3",
            r#"
fn foo() -> float {
    let x: float = 1.0
    for x: int8 = 1; x < 2; 1 {
        x
    }
    x
}
"#,
        ],
        [
            "antn_mismatch_1",
            r#"
fn foo() {
    for x: int8 = 1.0; x < 2; 1 {
        x
    }
}
"#,
        ],
        [
            "antn_mismatch_2",
            r#"
fn foo() {
    let y: int8 = 2
    for x: int = y; x < 2; 1 {
        x
    }
}
"#,
        ],
        [
            "cond_should_bool",
            r#"
fn foo() {
    for x: int = 1; 2.0; 1 {
        x
    }
}
"#,
        ],
        [
            "lit_step_mismatch",
            r#"
fn foo() {
    for x: int = 1; x < 3; 1.0 {
        x
    }
}
"#,
        ],
        [
            "var_step_mismatch",
            r#"
fn foo() {
    let y: int8 = 2
    for x: int = 1; x < 3; y {
        x
    }
}
"#,
        ],
    ];
    run_insta!("for", tests);
}

#[test]
fn test_ident() {
    let tests = [
        [
            "basic",
            r#"
fn foo(x: int) { x }
"#,
        ],
        [
            "unknown_var",
            r#"
fn foo(x: int) { y }
"#,
        ],
    ];
    run_insta!("ident", tests);
}

#[test]
fn test_let() {
    let tests = [
        [
            "basic",
            r#"
fn foo() {
    let x: int = 3
}
"#,
        ],
        [
            "basic_no_init",
            r#"
        fn foo() {
            let x: int
        }
        "#,
        ],
        [
            "lit_antn_mismatch",
            r#"
fn foo() {
    let x: int = 3.0
}
"#,
        ],
        [
            "var_antn_mismatch",
            r#"
fn foo() {
    let y: float = 1.0
    let x: int = y
}
"#,
        ],
        [
            "lit_range",
            r#"
fn foo() {
    let i: int = 2147483648
}
"#,
        ],
    ];
    run_insta!("let", tests);
}

#[test]
fn test_unary() {
    let tests = [
        [
            "lit",
            r#"
fn foo() { -3 }
"#,
        ],
        [
            "var",
            r#"
fn foo() {
    let xy: int8 = 1
    -xy
}
"#,
        ],
        [
            "bad_type_1",
            r#"
fn foo() {
    -false
}
"#,
        ],
        [
            "bad_type_2",
            r#"
fn foo() {
    -'c'
}
"#,
        ],
    ];
    run_insta!("unary", tests);
}

#[test]
fn test_array() {
    let tests = [
        ["type", "let x: [int; 3]"],
        ["lit", "let x: [int; 3] = [1, 2, 3]"],
        [
            "lit_reassign",
            r#"
let x: [int; 3] = [1, 2, 3]
x = [4, 5, 6]
"#,
        ],
        ["lit_bad_1", "let x: [int; 3] = [1, 2.0, 3]"],
        ["lit_bad_2", "let x: [float; 3] = [1, 2, 3]"],
        ["lit_bad_3", "let x: [int; 2] = [1, 2, 3]"],
        ["empty_lit", "let x: [int; 3] = []"],
        [
            "index_1",
            r#"
let x: [int; 3] = [1, 2, 3]
x[0]
"#,
        ],
        [
            "index_2",
            r#"
let x: [int; 3] = [1, 2, 3]
x[1 + 2]
"#,
        ],
        [
            "index_3",
            r#"
let x: [int; 3] = [1, 2, 3]
let y: int = 1
x[y]
"#,
        ],
        [
            "index_4",
            r#"
let x: [int; 3] = [1, 2, 3]
x[1] = 7
"#,
        ],
        [
            "index_bad_1",
            r#"
let x: [int; 3] = [1, 2, 3]
let y: float = 1.0
x[y]
"#,
        ],
        [
            "index_bad_2",
            r#"
let x: [int; 3] = [1, 2, 3]
x['c']
"#,
        ],
        [
            "index_bad_3",
            r#"
let x: [int; 3] = [1, 2, 3]
let y: int8 = 1
x[y]
"#,
        ],
    ];
    run_insta!("array", tests)
}

#[test]
fn test_scope() {
    let tests = [
        [
            "basic_shadowing",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let b: bool = false
    }
    b
}
"#,
        ],
        [
            "nested_shadowing",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let b: bool = false
        let a: float = {
            let b: float = 1.0
            b
        }
    }
    b
}
"#,
        ],
        [
            "delete_scope",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let c: int = 2
    }
    c
}
"#,
        ],
        [
            "for_scope",
            r#"
let x: float = 1.0
for x: int = 1; x < 10; 1 {
    x
}
x
"#,
        ],
        [
            "if_scope",
            r#"
let x: float = 1.0
if x < 2.0 {
    let y: int = 2
    x
}
y
"#,
        ],
        [
            "if_else_scope",
            r#"
let x: float = 1.0
if x < 2.0 {
    let y: int = 2
    x
} else {
    -y
}
"#,
        ],
    ];
    run_insta!("scope", tests);
}

#[test]
fn test_struct() {
    let tests = [
        [
            "fields_only",
            r#"
struct Foo {
    let a: int
    let b: float
}
"#,
        ],
        [
            "methods_only",
            r#"
struct Foo {
    fn c() -> int {
        1
    }
    fn d() {}
}
"#,
        ],
        [
            "mix",
            r#"
struct Foo {
    let a: int
    let b: float
    fn c() -> int { 1 }
}
"#,
        ],
        [
            "scope",
            r#"
fn main() {
    struct Foo {}
}
"#,
        ],
        [
            "field_selector",
            r#"
struct Foo {
    let a: int
}
struct Bar {
    let b: Foo
}
fn main() {
    let x: Foo
    let y: Bar
    x.a
    y.b
}
"#,
        ],
        [
            "field_selector_chained",
            r#"
struct Foo {
    let a: int
}
struct Bar {
    let b: Foo
}
fn main() {
    let x: Bar
    x.b.a
}
"#,
        ],
        [
            "struct_field_init",
            r#"
struct Foo {
    let a: int = 1
    fn bar() {}
}
fn main() {
    let f: Foo
    f.bar(1)
}
"#,
        ],
        [
            "field_selector_bad1",
            r#"
struct Foo {}
fn main() {
    let x: Foo
    x.a
}
"#,
        ],
        [
            "field_selector_bad2",
            r#"
struct Foo {
    let b: int
}
fn main() {
    let x: Foo
    x.a
}
"#,
        ],
        [
            "field_selector_bad3",
            r#"
struct Foo {
    let a: Bar
}
fn main() {
    let x: Foo
    x.a
}
"#,
        ],
        [
            "field_selector_bad4",
            r#"
fn main() {
    let x: int
    x.a
}
"#,
        ],
        [
            "method_selector",
            r#"
struct Foo {
    fn bar() {}
}
fn main() {
    let f: Foo
    f.bar()
}
"#,
        ],
        [
            "method_selector_bad1",
            r#"
fn main() {
    let f: int
    f.bar()
}
"#,
        ],
        [
            "method_selector_bad2",
            r#"
struct Foo {
    fn bar() {}
}
fn main() {
    let f: Foo
    f.baz()
}
"#,
        ],
        [
            "method_selector_bad3",
            r#"
struct Foo {
    fn bar() {}
}
fn main() {
    let f: Foo
    f.bar(1)
}
"#,
        ],
        ["field_init", "struct Foo { let a: int = 1 }"],
        [
            "self_type",
            r#"
struct Foo {
    fn bar() { self }
}
"#,
        ],
    ];
    run_insta!("struct", tests)
}

#[test]
fn test_invalid_types() {
    let tests = [
        [
            "let_good",
            r#"
struct Foo {}
fn main() {
    let x: int
    let y: Foo
    let z: Bar
}
struct Bar {}
"#,
        ],
        ["let_bad", "let x: blah"],
        [
            "for_bad",
            r#"
fn main() {
    for x: bad; x < 1; 1 { }
}
"#,
        ],
        ["proto_ret_bad", "fn main() -> Foo { }"],
        ["proto_arg_bad", "fn main(a: int, b: inti) { }"],
    ];
    run_insta!("invalid", tests)
}

#[test]
fn test_tych_int_no_hint() {
    use Literal::*;

    let literals = [
        (UInt64(7), Ok(Type::Int32)),
        (UInt64(i32::MAX as u64), Ok(Type::Int32)),
        (UInt64(i32::MAX as u64 + 1), Err(ERR_STR)),
        (Float(7.0), Ok(Type::Float)),
    ];

    let mut symbol_table = SymbolTable::new();
    let mut tc = Tych::new(&mut symbol_table);
    for lit in literals {
        let res = tc.visit_lit(lit.0, None).map(|e| e.ty().cloned().unwrap_or_default());
        assert_eq!(res, lit.1.map_err(|x| x.to_string()));
    }
}

#[test]
fn test_tych_int_with_hint() {
    use Literal::*;

    let literals = [
        (UInt64(7), Type::Int8, Ok(Type::Int8)),
        (UInt64(i8::MAX as u64), Type::Int8, Ok(Type::Int8)),
        (UInt64(i8::MAX as u64 + 1), Type::Int8, Err(ERR_STR)),
        (UInt64(7), Type::Int16, Ok(Type::Int16)),
        (UInt64(i16::MAX as u64), Type::Int16, Ok(Type::Int16)),
        (UInt64(i16::MAX as u64 + 1), Type::Int16, Err(ERR_STR)),
        (UInt64(7), Type::Int32, Ok(Type::Int32)),
        (UInt64(i32::MAX as u64), Type::Int32, Ok(Type::Int32)),
        (UInt64(i32::MAX as u64 + 1), Type::Int32, Err(ERR_STR)),
        (UInt64(7), Type::Int64, Ok(Type::Int64)),
        (UInt64(i64::MAX as u64), Type::Int64, Ok(Type::Int64)),
        (UInt64(i64::MAX as u64 + 1), Type::Int64, Err(ERR_STR)),
        (UInt64(7), Type::UInt8, Ok(Type::UInt8)),
        (UInt64(u8::MAX as u64), Type::UInt8, Ok(Type::UInt8)),
        (UInt64(u8::MAX as u64 + 1), Type::UInt8, Err(ERR_STR)),
        (UInt64(7), Type::UInt16, Ok(Type::UInt16)),
        (UInt64(u16::MAX as u64), Type::UInt16, Ok(Type::UInt16)),
        (UInt64(u16::MAX as u64 + 1), Type::UInt16, Err(ERR_STR)),
        (UInt64(7), Type::UInt32, Ok(Type::UInt32)),
        (UInt64(u32::MAX as u64), Type::UInt32, Ok(Type::UInt32)),
        (UInt64(u32::MAX as u64 + 1), Type::UInt32, Err(ERR_STR)),
        (UInt64(7), Type::UInt64, Ok(Type::UInt64)),
        (UInt64(u64::MAX as u64), Type::UInt64, Ok(Type::UInt64)),
        (Float(7.0), Type::Float, Ok(Type::Float)),
        (Float(7.0), Type::Double, Ok(Type::Double)),
    ];

    let mut symbol_table = SymbolTable::new();
    let mut tc = Tych::new(&mut symbol_table);
    for lit in literals {
        tc.hint = Some(lit.1);
        let res = tc.visit_lit(lit.0, None).map(|e| e.ty().cloned().unwrap_or_default());
        assert_eq!(res, lit.2.map_err(|x| x.to_string()));
    }
}

// let x: $variant
// x + 3
macro_rules! test_lit_hint_binop_int {
    ($variant:ident) => {{
        let mut st = SymbolTable::new();
        st.insert(Symbol::new_var("x", &$variant));
        let mut tc = Tych::new(&mut st);
        let lhs = ast::Node::new_ident(String::from("x"), None);
        let rhs = ast::Node::new_lit(Literal::UInt64(3), None);
        let res = tc.visit_binop(Operator::Add, lhs, rhs, None).map(|e| e.ty().unwrap_or_default().clone());
        assert_eq!(res, Ok($variant));

        let mut st = SymbolTable::new();
        st.insert(Symbol::new_var("x", &$variant));
        let mut tc = Tych::new(&mut st);
        let lhs = ast::Node::new_lit(Literal::UInt64(3), None);
        let rhs = ast::Node::new_ident(String::from("x"), None);
        let res = tc.visit_binop(Operator::Add, lhs, rhs, None).map(|e| e.ty().unwrap_or_default().clone());
        assert_eq!(res, Ok($variant));
    }};
}

// let x: $variant
// x + 3.0
macro_rules! test_lit_hint_binop_float {
    ($variant:ident) => {{
        let mut st = SymbolTable::new();
        st.insert(Symbol::new_var("x", &$variant));
        let mut tc = Tych::new(&mut st);
        let lhs = ast::Node::new_ident(String::from("x"), None);
        let rhs = ast::Node::new_lit(Literal::Float(3.0), None);
        let res = tc.visit_binop(Operator::Add, lhs, rhs, None).map(|e| e.ty().unwrap_or_default().clone());
        assert_eq!(res, Ok($variant));

        let mut st = SymbolTable::new();
        st.insert(Symbol::new_var("x", &$variant));
        let mut tc = Tych::new(&mut st);
        let lhs = ast::Node::new_lit(Literal::Float(3.0), None);
        let rhs = ast::Node::new_ident(String::from("x"), None);
        let res = tc.visit_binop(Operator::Add, lhs, rhs, None).map(|e| e.ty().unwrap_or_default().clone());
        assert_eq!(res, Ok($variant));
    }};
}

#[test]
fn test_tych_binop_lit() {
    use Type::*;

    test_lit_hint_binop_int!(Int8);
    test_lit_hint_binop_int!(Int16);
    test_lit_hint_binop_int!(Int32);
    test_lit_hint_binop_int!(Int64);
    test_lit_hint_binop_int!(UInt8);
    test_lit_hint_binop_int!(UInt16);
    test_lit_hint_binop_int!(UInt32);
    test_lit_hint_binop_int!(UInt64);
    test_lit_hint_binop_float!(Float);
    test_lit_hint_binop_float!(Double);
}
