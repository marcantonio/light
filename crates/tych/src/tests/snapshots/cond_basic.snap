---
source: crates/tych/src/tests.rs
assertion_line: 291
expression: "(test[1], res)"
---
- "\nfn foo() {\n    if 1 > 2 {\n        1\n    } else {\n        2\n    }\n}\n"
- Ok:
    nodes:
      - kind:
          Fn:
            proto:
              name: "main::foo"
              args: []
              ret_ty: Void
              is_extern: false
              module: main
            body:
              kind:
                Block:
                  list:
                    - kind:
                        Cond:
                          cond_expr:
                            kind:
                              BinOp:
                                op: Gt
                                lhs:
                                  kind:
                                    Lit:
                                      value:
                                        Int32: 1
                                      ty: Int32
                                rhs:
                                  kind:
                                    Lit:
                                      value:
                                        Int32: 2
                                      ty: Int32
                                ty: Bool
                          then_block:
                            kind:
                              Block:
                                list:
                                  - kind:
                                      Lit:
                                        value:
                                          Int32: 1
                                        ty: Int32
                                ty: Int32
                          else_block:
                            kind:
                              Block:
                                list:
                                  - kind:
                                      Lit:
                                        value:
                                          Int32: 2
                                        ty: Int32
                                ty: Int32
                          ty: Int32
                  ty: Int32

